pageserver/src/walreceiver.rs

```diff
@@ -64,375 +80,1133 @@ thread_local! {
     pub(crate) static IS_WAL_RECEIVER: Cell<bool> = Cell::new(false);
 }
 
-fn drop_wal_receiver(tenantid: ZTenantId, timelineid: ZTimelineId) {
-    let mut receivers = WAL_RECEIVERS.lock().unwrap();
-    receivers.remove(&(tenantid, timelineid));
+/// WAL receiver state for sharing with the outside world.
+/// Only entries for timelines currently available in pageserver are stored.
+static WAL_RECEIVER_ENTRIES: Lazy<RwLock<HashMap<ZTenantTimelineId, WalReceiverEntry>>> =
+    Lazy::new(|| RwLock::new(HashMap::new()));
+
+/// Gets the public WAL streaming entry for a certain timeline.
+pub async fn get_wal_receiver_entry(
+    tenant_id: ZTenantId,
+    timeline_id: ZTimelineId,
+) -> Option<WalReceiverEntry> {
+    WAL_RECEIVER_ENTRIES
+        .read()
+        .await
+        .get(&ZTenantTimelineId::new(tenant_id, timeline_id))
+        .cloned()
 }
 
-// Launch a new WAL receiver, or tell one that's running about change in connection string
-pub fn launch_wal_receiver(
+/// Sets up the main WAL receiver thread that manages the rest of the subtasks inside of it, per timeline.
+/// See comments in [`wal_receiver_main_thread_loop_step`] for more details on per timeline activities.
+pub fn init_wal_receiver_main_thread(
     conf: &'static PageServerConf,
-    tenantid: ZTenantId,
-    timelineid: ZTimelineId,
-    wal_producer_connstr: &str,
-) -> Result<()> {
-    let mut receivers = WAL_RECEIVERS.lock().unwrap();
-
-    match receivers.get_mut(&(tenantid, timelineid)) {
-        Some(receiver) => {
-            debug!("wal receiver already running, updating connection string");
-            receiver.wal_producer_connstr = wal_producer_connstr.into();
+    mut timeline_updates_receiver: mpsc::UnboundedReceiver<LocalTimelineUpdate>,
+) -> anyhow::Result<()> {
+    let etcd_endpoints = conf.broker_endpoints.clone();
+    ensure!(
+        !etcd_endpoints.is_empty(),
+        "Cannot start wal receiver: etcd endpoints are empty"
+    );
+    let broker_prefix = &conf.broker_etcd_prefix;
+    info!(
+        "Starting wal receiver main thread, etdc endpoints: {}",
+        etcd_endpoints.iter().map(Url::to_string).join(", ")
+    );
+
+    let runtime = tokio::runtime::Builder::new_multi_thread()
+        .thread_name("wal-receiver-runtime-thread")
+        .worker_threads(40)
```

Comment:
Don't think this requires immediate action, but I'm still dubious about running walreceivers in general tokio runtime. I'd expect them to be not idle-waiting-for-network most of the time, in which case async runtime is pointless. Moreover, operations like locking layer while writing to it are not async friendly and will block the runtime with etcd communcation and main wal receiver activity.

Reply:
>  in which case async runtime is pointless.

That's simply not true. Async is useful not only in places where most of the tasks are idle. E g for cancellation purposes we will need to use per-thread runtimes in each walreceiver which IMO is not ideal because one of the main reasons for introducing async paradigm was to decrease the overhead for the separate execution thread so one can have thousands of coroutines running concurrently (and some of them simultaneously). Per thread runtimes make sense when you want to reduce cross core coordination and increase thread locality. It works via spawning threads for every available processor core and using a separate runtime in every thread. This is the paradigm that is in use by [Actix](https://actix.rs/) or nginx. Secondly, aside from cancellation semantics async is capable of introducing higher parallelization via adding more concurrency to io operations. E g you can schedule many write operations simultaneously and wait for them at once instead of issuing them one by one. Not to mention io_uring capabilities.

Also, there are some attempts to reduce thread count: https://github.com/neondatabase/neon/pull/1794, and IMO having per thread wal_receivers contradicts this direction.

Also, there is nothing wrong with using async for more CPU-intensive tasks. See the article: https://www.influxdata.com/blog/using-rustlangs-async-tokio-runtime-for-cpu-bound-tasks/ and write up on the subject in the repository https://github.com/influxdata/influxdb_iox/blob/main/docs/multi_core_tasks.md

Blocking is a problem and it partially can be solved by spawn_blocking, but it uses threadpool under the hood so a bit of a compromise.

To be clear, I do not insist on rewriting everything to async. It has its own problems. But let's be fair in accounting for its benefits too.